### ISSUE TITLE
Inconsistent environment variables (e.g., LANG) between built-in terminal and external terminal when using Fish shell in VSCode.

### ISSUE THREAD

**Issue by cbconne** [microsoft/vscode#182938]
Type: <b>Bug</b>

Different results observed when executing echo $LANG in the built-in terminal of VSCode and an external terminal (iTerm2).
![image](https://github.com/microsoft/vscode/assets/44186296/68f4b883-de06-4209-a5c9-6ec88d0e8c1d)

VS Code version: Code 1.78.2 (b3e4e68a0bc097f0ae7907b217c1119af9e03435, 2023-05-10T15:00:40.428Z)
OS version: Darwin arm64 22.4.0
Modes:
Sandboxed: Yes

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M1 (8 x 24)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|1, 2, 2|
|Memory (System)|16.00GB (0.06GB free)|
|Process Argv|--crash-reporter-id 243db59d-6445-4d8c-b90f-6c47d6b18e81|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (52)</summary>

Extension|Author (truncated)|Version
---|---|---
jsoncrack-vscode|Ayk|1.0.8
markdown-mermaid|bie|1.18.1
path-intellisense|chr|2.8.4
vscode-eslint|dba|2.4.0
gitlens|eam|13.6.0
vscode-great-icons|emm|2.1.93
prettier-vscode|esb|9.12.0
copilot|Git|1.86.82
go|gol|0.38.0
fastcomplete|How|2.0.0
comment-translate|int|2.2.4
jenkinsfile-support|ivo|1.1.0
vscode-edit-csv|jan|0.7.6
jenkins-pipeline-linter-connector|jan|1.2.0
plantuml|jeb|2.17.5
tencent-cloud-translate|Kai|0.0.3
inifmt|lkr|0.1.5
git-graph|mhu|1.30.0
vscode-docker|ms-|1.25.1
flake8|ms-|2023.4.0
isort|ms-|2022.8.0
python|ms-|2023.8.0
vscode-pylance|ms-|2023.5.30
jupyter|ms-|2023.4.1011241018
jupyter-keymap|ms-|1.1.0
jupyter-renderers|ms-|1.0.15
vscode-jupyter-cell-tags|ms-|0.1.8
vscode-jupyter-slideshow|ms-|0.1.5
remote-ssh|ms-|0.102.0
remote-ssh-edit|ms-|0.86.0
live-server|ms-|0.4.8
remote-explorer|ms-|0.4.0
material-icon-theme|PKi|4.27.0
vscode-xml|red|0.25.0
vscode-yaml|red|1.12.2
vue-vscode-snippets|sdr|3.1.1
markdown-preview-enhanced|shd|0.6.8
vscode-counter|uct|3.2.1
errorlens|use|3.11.0
gitflow|vec|1.2.1
intellicode-api-usage-examples|Vis|0.2.7
vscodeintellicode|Vis|1.2.30
vscode-icons|vsc|12.4.0
vim|vsc|1.25.2
volar|Vue|1.6.4
vscode-typescript-vue-plugin|Vue|1.6.5
change-case|wma|1.0.0
Pangu-Markdown|xlt|0.0.6
markdown-pdf|yza|1.4.4
markdown-all-in-one|yzh|3.5.1
json|Zai|2.0.2
material-theme|zhu|3.15.13

(4 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
vserr242cf:30382550
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscod805:30301674
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
cmake_vspar411:30581797
vsaa593cf:30376535
pythonvs932:30410667
cppdebug:30492333
vscaat:30438848
vsclangdf:30486550
c4g48928:30535728
dsvsc012:30540252
pynewext54:30695312
azure-dev_surveyone:30548225
vscccc:30610679
282f8724:30602487
pyind779:30671433
89544117:30613380
pythonsymbol12:30671437
a9j8j154:30646983
showlangstatbar:30737416
vsctsb:30705553
pythonms35:30701012
pythonfmttext:30731395
pythoncmv:30739700
fixshowwlkth:30730052
pythongtdpath:30739704

```

</details>

**Comment by meganrogge** [comment]
Thanks for creating this issue! To help narrow it down, please take the time to follow below steps. Thanks a lot and happy coding 🙏 

### Is the issue caused by an Extension?
Stop VS Code and from the command line (NOT the integrated terminal in Code), execute: `code --disable-extensions` and try your steps again to see if it reproduces. If it still reproduces, jump to the next chapter.

If you see it is an issue with the extension, please run the command `Start Extension Bisect` and follow the instructions to find the extension that is causing this issue.

<img width="631" alt="image" src="https://user-images.githubusercontent.com/900690/228760008-d5790ad1-74da-46a2-917f-e1bd26281ca8.png">

Please report the issue to the extension causing this.

### Is the issue caused by VS Code Core?
If the issue is not caused by an extension, it is maybe caused by your configuration. Try to run `code --disable-extensions --user-data-dir <directory>` where `<directory>` is an empty folder. This will ensure Code is starting with a fresh data directory, e.g. no specific settings and without any extensions running. 

If it still reproduces, it was maybe fixed already in our nightly release. Please download from https://code.visualstudio.com/insiders/ and see if you can reproduce the issue.

Finally, to help us narrow down what change caused the issue, please run the following command to find the build. It will start previous releases of VS Code insiders to see where the regression was introduced:
* `npx --yes vscode-bisect@latest`
* follow the instructions until you found the offending build
* report back the commit range in this issue

Thanks!

**Comment by cbconne** [comment]
I tried running code --disable-extensions --user-data-dir <directory> and it worked. Then, I noticed that running code directly from the terminal also works.

In summary, when I launch VSCode by entering "code" in the terminal, the result of echo $LANG is consistent with my external terminal. However, when I launch VSCode using the "Visual Studio Code.app" in the "Applications" folder, the result of echo $LANG is inconsistent.

Furthermore, I discovered that this issue is not specific to using Fish or Zsh, so I modified the title of the issue accordingly.

1. Launching from the command line:

<img width="825" alt="image" src="https://github.com/microsoft/vscode/assets/44186296/a6b465bc-7e55-4f58-8c6c-95a2724629d3">

2. Launching from the "Applications":

<img width="739" alt="image" src="https://github.com/microsoft/vscode/assets/44186296/c161c10c-7a9d-4073-8695-5a43f687c294">

And I did not use the machine I used to submit the issue for troubleshooting. Instead, I used another MacBook Air with the same versions of VSCode and macOS (and encountered the same problem).

**Comment by cbconne** [comment]
nightly release have the same problem: I launch VSCode using the "Visual Studio Code - Insiders.app"
<img width="686" alt="image" src="https://github.com/microsoft/vscode/assets/44186296/3aaca6a8-f28e-4153-a8f5-7d4cc47e7599">

**Comment by cbconne** [comment]
https://code.visualstudio.com/docs/supporting/faq#_resolving-shell-environment-fails

To find the problem I uninstalled fish, tried using zsh to find the problem. Then found some new error messages:

<img width="395" alt="image" src="https://github.com/microsoft/vscode/assets/44186296/5df715fb-a0fa-42ce-8ea4-8b5846d91566">

Following the prompts of VSCode, I saw such a description in the official FAQ

<img width="780" alt="image" src="https://github.com/microsoft/vscode/assets/44186296/f3410cb4-074c-40db-b9a1-3c91f94323df">

I think I've found the reason: it looks like the "small process to run (or "resolve") the shell environment" isn't working - or it's reading the wrong .xshrc info, I can't be sure.

**Comment by Tyriar** [comment]
Yes it's probably due to the process environment. Additionally there is the detectLocale setting which can influence the value of `$LANG`: https://code.visualstudio.com/docs/terminal/advanced#_interaction-with-lang

**Comment by cbconne** [comment]
> Yes it's probably due to the process environment. Additionally there is the detectLocale setting which can influence the value of `$LANG`: https://code.visualstudio.com/docs/terminal/advanced#_interaction-with-lang

Oh, I finally know where this "en-US.UTF-8" comes from, thanks!
I don't `export LANG` in my zshrc, but macOS (I guess) has `LANG=zh_CN.UTF-8` set in my terminal environment variables based on my chosen language and region. When I open VSCode from the terminal through `code`, it inherits this environment variable, so terminal.integrated.detectLocale does not take effect; but when I open it through the application, it does not have this environment variable, so the setting terminal.integrated.detectLocale set $LANG to "en-US.UTF-8".
Thank you both! @meganrogge @Tyriar
